### PR TITLE
qml.equal for ParametrizedEvolution

### PR DESCRIPTION
**Pull request by lillian542**
**Context:**
If two ParametrizedEvolution operations differ by their time windows, coefficients or base operators, they should be recognized to be different, but they aren't.

**Description of the Change:**
Single dispatch function added for ParametrizedEvolution operators.

**Benefits:**
qml.equal no longer identifies different pulses as equal.

**Related GitHub Issues:**
#3846

### CODE REVIEW

File: pennylane/ops/functions/equal.py [hunk 1]
```diff
@@ -302,6 +303,24 @@ def _equal_hamiltonian(op1: Hamiltonian, op2: Observable, **kwargs):
...
+    # check parameters passed to operator match
+    if not _equal_operators(op1, op2, **kwargs):
+        return False
+
+    # check H.coeffs match
+    if not all(c1 == c2 for c1, c2 in zip(op1.H.coeffs, op2.H.coeffs)):
+        return False
+
+    # check that all the base operators on op1.H and op2.H match
+    return all(equal(o1, o2, **kwargs) for o1, o2 in zip(op1.H.ops, op2.H.ops))
```
Comment (dwierichs) [inline pennylane/ops/functions/equal.py:321]: Can't we use `_equal_operator` here as well? :thinking:

File: tests/ops/functions/test_equal.py [hunk 2]
```diff
@@ -1568,3 +1568,99 @@ def test_sum_with_multi_wire_operations(self, base_list1, base_list2, res):
...
+
+
+f1 = qml.pulse.pwc([0.3, 1.0])
+
+
+def f2(p, t):
+    return p[0] * t**2 + p[1]
+
+
+class TestParametrizedEvolutionComparisons:
```
Comment (dwierichs) [inline tests/ops/functions/test_equal.py:1580]: Could you add docstrings to the class and functions? :)

**Review by dwierichs — COMMENTED** [review]
Nicely done, very clear modular testing! :100:

Don't have anything major, just a test-docstring comment and a question.

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -29,6 +29,9 @@
   to be set as trainable), and therefore the derivatives are computed more efficiently.
   [(#3697)](https://github.com/PennyLaneAI/pennylane/pull/3697)
 
+* The `qml.equal` function now handles comparisons of `ParametrizedEvolution` operators.
```
Comment (mudit2812) [inline doc/releases/changelog-dev.md:32]: Not sure if this should be in the improvements or bug fixes section.

**Review by mudit2812 — COMMENTED** [review]
Looks good. I think David asked all the questions I was wondering. Once they are addressed, I'll be happy to approve 🚀

File: doc/releases/changelog-dev.md (hunk 3 above)
Comment (lillian542) [reply doc/releases/changelog-dev.md:32]: Good point, bug fixes might make more sense!

File: pennylane/ops/functions/equal.py (hunk 1 above)
Comment (lillian542) [reply pennylane/ops/functions/equal.py:321]: No - this should dispatch to the specific relevant function. The `_equal_operator` function is a fallback for operators that don't have a more specific function to handle them. All the `SymbolicOp` and `CompositeOp` operators need to go to their own functions. 

What if one of the operators the ParametrizedHamiltonian is `SProd`, for example?

File: tests/ops/functions/test_equal.py (hunk 2 above)
Comment (lillian542) [reply tests/ops/functions/test_equal.py:1580]: Added :)

Comment (dwierichs) [reply tests/ops/functions/test_equal.py:1580]: Thanks :100:

**Comment by albi3ro** [comment]
[sc-34408]

**Review by dwierichs — APPROVED** [review]
Looks good to me, thanks ! :)

**Review by mudit2812 — APPROVED** [review]
Thanks for the changes Lillian 🚀 ! I think this is okay to merge. One use case that is missed is if someone creates 2 identical callables that are used as parameters for a `ParametrizedEvolution`:

```python
f1 = lambda p, t: p ** t
f2 = lambda p, t: p ** t

H1 = qml.dot([f1], [qml.PauliX(0)])
H2 = qml.dot([f2], [qml.PauliX(0)])

ev1 = qml.evolve(H1)
ev2 = qml.evolve(H2)
```
```pycon
>>> params = [0.4]
>>> t = 5
>>> qml.equal(ev1(params, t), ev2(params, t))
False
```
Note that this only fails when you create two callable parameters _from scratch_ that perform the same computation. If you copy a callable (shallow or deep copy), `qml.equal` works correctly. This is an **extreme** edge case and I'm not even sure if it can be addressed. I'm fine with merging without addressing it - no user should use `ParametrizedEvolution` this way anyway.

**Pull request merged**